integrations/ad_integration/ad_sync.py

```diff
@@ -346,24 +365,108 @@ def _edit_user_addresses(self, uuid, ad_object):
                 self._create_address(uuid, ad_object[field], klasse)
             else:
                 # This is an existing address
-                if not fields_to_edit[field][1] == ad_object[field]:
+                if not fields_to_edit[field]['value'] == ad_object[field]:
                     msg = 'Value change, MO: {} <> AD: {}'
                     self.stats['addresses'][1] += 1
                     self.stats['users'].add(uuid)
-                    self._edit_address(fields_to_edit[field][0],
+                    self._edit_address(fields_to_edit[field]['uuid'],
                                        ad_object[field],
                                        klasse)
                 else:
                     msg = 'No value change: {}=={}'
-                logger.debug(msg.format(fields_to_edit[field][1],
+                logger.debug(msg.format(fields_to_edit[field]['value'],
                                         ad_object[field]))
 
+    def _finalize_it_system(self, uuid, ad_object):
+        if 'it_systems' not in self.mapping:
+            return
+
+        today = datetime.strftime(datetime.now(), "%Y-%m-%d")
+        it_systems = {
+            itsystem['uuid']: itsystem for itsystem in
+                self.helper._mo_lookup(uuid, 'e/{}/details/it')
+        }
+
+        def check_validity_is_ok(uuid):
+            # NOTE: Maybe this should be not set, or in the future?
+            return it_systems[uuid]['validity']['to'] is None
+
+        # Find fields to terminate
+        it_system_uuids = self.mapping['it_systems'].values()
+        it_system_uuids = filter(check_validity_is_ok, it_system_uuids)
+
+        for uuid in it_system_uuids:
+            payload = {
+                'type': 'it',
+                'uuid': uuid,
+                'validity': {"to": today}
+            }
+            logger.debug('Finalize payload: {}'.format(payload))
+            response = self.helper._mo_post('details/terminate', payload)
+            logger.debug('Response: {}'.format(response.text))
+
+    def _finalize_user_addresses(self, uuid, ad_object):
+        if 'user_addresses' not in self.mapping:
+            return
+
+        today = datetime.strftime(datetime.now(), "%Y-%m-%d")
+        fields_to_edit = self._find_existing_ad_address_types(uuid)
+
+        def check_ad_field_exists(field):
+            if field not in ad_object:
+                logger.debug('No such AD field: {}'.format(field))
+                return False
+            return True
+
+        def check_field_in_fields_to_edit(field):
+            return field in fields_to_edit.keys()
+
+        def check_validity_is_ok(field):
+            # NOTE: Maybe this should be not set, or in the future?
+            return fields_to_edit[field]['validity']['to'] is None
+
+        # Find fields to terminate
```

Comment:
Er det her fordi du kun vil terminere properties for brugeren, hvis disse properties kommer fra AD? Sådan at manuelt oprettede settings får lov at leve videre?

Reply:
Yes, det tror jeg var tanken